### PR TITLE
feat(staleness): nightly pass, volatility tags, and recall warnings (#228)

### LINKED ISSUE #228: Time-triggered staleness pass: volatility-gated flagging + compound-staleness guard at recall
> **Scope updated 2026-07-28** after Product Hunt v2.1 feedback. See [Revised scope](#revised-scope-2026-07-28-ph-v21-feedback) below — age is now the *trigger*, not the flag, and compound staleness is in scope. Some original acceptance criteria are superseded.
>
> Code references below were repointed after the v2.1 modular-Worker split (`src/index.ts` is no longer the monolith).

## Feedback (raised across two launches)

Christopher (this launch) and Kyo (v2). Two distinct failure modes; only one is handled today:

- **Wrong from birth** (Gal Dayan, v2): a memory was never true. No competing write ever arrives, so draft-vs-canonical never fires.
- **Silent staleness** (Christopher, Kyo): it was true when written and reality moved. Same problem — no conflict signal, so a canonical entry sits there being confidently wrong forever.

## Why today's system misses it

The conflict/contradiction system is **write-triggered only** — see the comment at `src/recall/math.ts:21`: *"Staleness is handled by status + contradiction, not by making old memories invisible."* And canonical entries get `RECENCY_FLOOR_DURABLE = 0.9` (`src/recall/math.ts:24`, applied at `src/recall/math.ts:71-74`), so they barely decay and stay top-ranked indefinitely. Nothing time-triggered ever re-examines an aged canonical.

## Proposed change (new nightly job, existing infra)

The scheduled handler already fans out three bounded jobs — `runNightlyCompression` + `runGraphPass` + `runScheduledIntegrationSync` (`src/index.ts:42-45`). Add a fourth: a **staleness pass**.

- **MVP:** flag canonicals untouched past an age threshold and surface an age qualifier at recall — e.g. *"true as of &lt;date&gt;, verify before asserting"* — so the AI treats it as point-in-time rather than fact. (This was Kyo's own suggestion at v2.)
- **Fast-follow:** prompt for re-verification of flagged entries.

Uses the reserved-tag mechanism (like `status:`) — no schema change.

---

## Revised scope (2026-07-28, PH v2.1 feedback)

Three commenters on the v2.1 Product Hunt post independently hit the same limitation: the new "why it appeared" recall provenance explains **retrieval**, not **truth**. A memory can be perfectly connected and six months out of date, and the connection view looks identical either way. Two of their points change what this issue needs to do.

### 1. Age is the trigger, volatility is the flag

*Raised by Jernej Jan Kočica.*

Age and staleness are not the same thing. Some memories never expire — a name, a birthday, where someone grew up. Flag those by age and the qualifier becomes noise users learn to skip, which destroys the value of the qualifier everywhere else. The memories worth flagging are the ones about something that can move underneath you: employment status, address, a plan, which tool someone uses. Proposed cheap heuristic: does the memory describe a **state** or a **fact**.

**Why this changes the scope as filed:** this issue keys the flag off age. Age alone will over-flag durable memories and train users to ignore the signal.

**What already exists — do not re-propose as new.** The ranking layer already has a three-tier durability axis, shipped in #213 (recall quality work, July 2026), now at `src/recall/math.ts:23-25`:

| Constant | Value | Applied when |
| --- | --- | --- |
| `RECENCY_FLOOR` | 0.6 | ordinary |
| `RECENCY_FLOOR_DURABLE` | 0.9 | `getStatus(tags) === 'canonical'` OR `importance >= 4` |
| `RECENCY_FLOOR_VOLATILE` | 0.15 | `task` tag |

So different memories already decay at different rates. **The weakness is the input, not the mechanism.** Those tiers are fed by *proxies* for volatility (canonical status, importance score, the `task` tag), not by volatility itself. A birthday gets the durable floor only because someone marked it important, not because the system understands that birthdays do not move.

**Precision note — `kind:` is a different axis and cannot be reused here.** The existing episodic/semantic classifier (`kind:` reserved tag namespace, shipped v1.9 in #148, now `src/memory/kind.ts`) measures *event vs fact*. Volatility measures *moving vs fixed*. These cross: someone's employer is a semantic **fact** and simultaneously the thing most likely to be wrong in a year. Treating `kind:semantic` as a trust signal would misclassify exactly the memories most at risk.

**Design conclusion:** age is not the flag, it is the trigger. Age says *re-check this*. Volatility says *whether re-checking is worth interrupting the user for*. A volatility classifier — or the cheap state-vs-fact proxy — is the missing input, and the existing recency-floor tiering at `src/recall/math.ts:71-74` is where it should feed.

### 2. Compound staleness

*Raised by Omri Ben-Shoham.*

The risky case is not one stale memory. It is two stale-but-plausible memories getting connected into a synthesis that is wrong in a way neither source alone would be. An old calendar entry plus an old email can combine into a confident-sounding "fact" that never existed on its own. The why-it-appeared view would look completely clean in that case, because each piece really did connect logically. The failure is that both inputs expired quietly.

**Why the current design misses it:** staleness today is handled by status plus contradiction detection, by explicit design (`src/recall/math.ts:21`). Contradiction detection is **write-triggered only**. Two aged entries combining at **read** time never trip it, because nothing new was written to contradict.

**What this changes about the requirement:** it shifts from "flag old entries" to "flag syntheses built from old entries." That is a different fix and probably a different code path — it lives at the multi-hop graph expansion / recall assembly layer (`expandGraph`, `src/graph/traverse.ts:29`, called from `src/recall/search.ts:188`) rather than in per-entry scoring.

**Implementation trap worth recording:** the obvious home for a synthesis-level guard is `synthesizeInsight` (`src/recall/insight.ts:5`), but it is gated on the `synthesize` flag consumed at `src/recall/search.ts:265`, and the MCP path opts out explicitly — `synthesize: false` at `src/mcp/server.ts:202` (#219). A guard placed only there would miss the assistant path entirely — which is precisely where the compound-staleness failure Omri describes actually lands, since the assistant does its own synthesizing from the returned matches.

**And re-enabling synthesis on MCP is not the fix.** `synthesizeInsight` is a Workers AI inference call (`LLM_MODEL` via `readStreamText`), so #219 turned it off on that path to cut **neuron usage** — a cost/quota constraint, not a latency tradeoff that can be dialled back for this feature. Every MCP recall would burn an inference just to attach a warning. The staleness signal therefore has to travel with the recall payload as data, computed without an LLM call, not be embedded in a server-side insight string.

### Context raising the priority

- **Nightly cron infra already exists**, so this is a new job and not new infra (`src/index.ts:42-45`).
- **Write dates already ship end to end.** Every memory stores its written date, it travels with recall output into the assistant (not dashboard-only), and the connection/provenance view displays it inline. Confirmed 2026-07-28. This issue does not need to add date surfacing, only interpretation of it.
- **Calendar and email capture (new in v2.1) raises the stakes.** Both are full of things that were true when written and never corrected: the meeting that moved, the address that changed. Two commenters raised this independently as the reason to prioritise this issue.

---

## Acceptance

Still holds:

- Recall output includes an "as of" qualifier for flagged entries.
- The new job is bounded like the other nightly jobs.

~~Canonicals aged past the threshold carry a staleness flag.~~ **Superseded 2026-07-28** — assumes age-only flagging. Replaced by:

- Age past threshold **triggers** re-examination; a volatility signal (state-vs-fact classification, or an explicit volatility tier) decides whether the entry is actually flagged. Durable canonicals crossing the age threshold must **not** be flagged.
- The volatility signal feeds the existing recency-floor tiering (`src/recall/math.ts:71-74`) rather than introducing a parallel ranking mechanism.
- `kind:` (episodic/semantic) is **not** used as the volatility input.

Added:

- A recall result assembled from multiple long-untouched inputs carries a synthesis-level staleness signal, even when every individual link is valid.
- That signal is present on the MCP/assistant path, not only in the server-side insight string.
- The signal is computed **without an extra Workers AI inference call**, so it does not reintroduce the neuron cost #219 removed from the MCP path.

### PR DESCRIPTION
**Pull request by FaberVi**
## Summary
- Add nightly staleness pass (90-day age gate, 25-entry bound) with heuristic volatility classification (`volatility:*`, `stale:as-of`)
- Track `updated_at` on writes; clear `stale:as-of` on append/merge/replace/update
- Surface per-match and compound staleness in HTTP recall JSON, MCP recall text, and dashboard
- Use volatility tags for recency floors in ranking; protect reserved tags from compression
- Regression tests for pass edge cases, write-path clears, recall JSON/MCP output, compress reserved tags, and volatile floor override

## Scope
Single commit on top of `main` — staleness only (#228). Unrelated refactor/i18n/#223 commits removed from this branch.

## Test plan
- [x] `npm run typecheck`
- [x] `npm test`
- [ ] Verify nightly cron flags old state/volatile memories and skips durable ones
- [ ] Confirm append/merge/replace clears `stale:as-of` and bumps `updated_at`
- [ ] Check GET `/recall` returns `stale_as_of`, `updated_at`, `compound_stale`
- [ ] Check MCP `recall` tool shows staleness warning for 2+ stale matches

Closes #228

### CODE REVIEW

**Review by rahilp — CHANGES_REQUESTED** [review]
Thanks for picking this one up — the shape is right. The nightly job slots into the existing cron fan-out cleanly, routing volatility into `getRecencyFloor` instead of building a parallel ranking mechanism is exactly what the issue asked for, and protecting the new reserved tags from compression (`digest.ts`, `nightly.ts`) is the kind of detail that's easy to miss.

I can't take it as-is though. I ran the pass and the classifier against realistic data rather than reading them, and found four blockers. CI being green doesn't clear them — I'll explain at the end why no test in the repo can currently catch two of these.

---

## 1. The pass never gets past the first 25 entries

`src/staleness/pass.ts` selects:

```sql
ORDER BY COALESCE(updated_at, created_at) ASC LIMIT 25
```

The `UPDATE` at the end of the loop writes only `tags` — it never bumps `updated_at`, and nothing marks a row as processed. So the query returns the **same 25 oldest rows every night**.

Running the pass 10 times over a DB with 60 aged entries:

```
flagged after each of 10 nights: [25,25,25,25,25,25,25,25,25,25]
```

Entries 26+ are never examined. On any brain with more than 25 aged memories the feature does nothing after night one.

`test/unit/staleness-pass.test.ts` asserts the 25-row bound but only ever calls `runStalenessPass` once, so the bound looks correct and the non-convergence is invisible. A second invocation in that test is the assertion that catches this.

Worth flagging that the obvious patch doesn't work: adding `AND tags NOT LIKE '%volatility:%'` still loops forever, because `classifyVolatility` returns `null` for most content and those rows never get a volatility tag. This needs a real processed marker or a cursor — your call which, but it's a design decision, not a one-liner.

## 2. `updated_at NOT NULL` gives new and existing brains different schemas

`src/db/init.ts` adds the column two different ways:

```sql
CREATE TABLE IF NOT EXISTS entries (..., updated_at INTEGER NOT NULL, ...)   -- new brains
ALTER TABLE entries ADD COLUMN updated_at INTEGER                            -- existing brains (nullable)
```

Replaying the pre-PR 6-column INSERT against both, in sqlite:

```
fresh brain    → Runtime error: NOT NULL constraint failed: entries.updated_at
existing brain → inserted ok, updated_at = NULL
```

The consequence is a rollback trap: ship this, create a brain, then roll the Worker back to `main` for any reason, and capture is permanently broken on that brain — `main`'s INSERT doesn't supply `updated_at` and the column has no default.

Fix is either `updated_at INTEGER NOT NULL DEFAULT 0` in the `CREATE TABLE`, or leave it out of `CREATE TABLE` entirely and let the `ALTER` own it on both paths. The second keeps one schema everywhere, which I'd prefer.

Related, lower priority: the backfill `UPDATE entries SET updated_at = created_at WHERE updated_at IS NULL` runs on every `initializeDatabase` call and `updated_at` isn't indexed, so it's an unindexed scan on every cold isolate, forever. Worth a guard.

## 3. This silently blocks #264 — and #264 is already broken without it

`src/entries/import.ts` in #264 inserts `updated_at`, but #264 never touches `src/db/init.ts`. Against `main`'s schema:

```
Parse error: table entries has no column named updated_at
```

So `POST /import` fails on every brain today, and #264 has an undeclared hard dependency on this PR. Please either note the ordering on #264 or move the column migration there — but the two need to land in a known order.

## 4. The classifier silently re-ranks memories the user never touched

Two separate problems that compound.

**The patterns are too broad.** `/\buses?\s+\w+/i`, `/\baddress\b/i`, `/\bschedule[ds]?\b/i` and `/\bcurrently\b/i` match ordinary prose. Over 20 memories drawn from this project's own stored content, 3 of 14 neutral notes were flagged:

```
FLAG  state     The user's email address is rahil@piranifam.com.
FLAG  volatile  The scheduled handler fans out four bounded nightly jobs.
FLAG  state     Cloudflare's API uses PUT /accounts/{id}/workers/scripts/{s}/secrets.
```

An email address is close to the archetypal *durable* fact and it gets "verify before asserting." This is the specific failure the revised issue scope warns about: over-flagging trains users to skip the qualifier, which destroys its value everywhere else.

**Volatility outranks canonical and importance.** `getRecencyFloor` in `src/recall/math.ts` checks `getVolatility` before `status:canonical` / `imp >= 4`, so the regex wins. Measured:

| tags | floor |
| --- | --- |
| `status:canonical` | 0.90 |
| `status:canonical` + `volatility:state` | 0.60 |
| `status:canonical` + `volatility:volatile` | **0.15** |
| importance 5 + `volatility:volatile` | **0.15** |

A canonical memory containing the word "scheduled" has its recency floor cut 6×, retroactively, by a cron job, with no user action and no visible cause.

Two asks here:

- Make `volatility:` a **fallback below** the existing canonical/importance proxies rather than an override. An explicit user signal should beat a heuristic.
- `withVolatility` replaces unconditionally, so the pass overwrites a manually-set `volatility:` tag every night — and because of issue 1 it's the same 25 rows being overwritten repeatedly. Once a human has set it, the pass should leave it alone.

Also worth noting the classifier is English-only, which sits awkwardly next to the EN/IT installer work. Not a blocker — it degrades to `null`, which is the safe direction — but it means the feature is a no-op for non-English memories and that should be stated somewhere.

---

## Compound staleness doesn't do what the issue asked

`computeCompoundStale` counts matches that already carry `stale:as-of`, so it's entirely derivative of the per-entry flag: it can't fire until the nightly pass has flagged two of them, which given issue 1 is essentially never on a real brain.

The acceptance criterion is a signal on *a recall result assembled from multiple long-untouched inputs*. That can be computed directly from each match's `updatedAt` at assembly time — no cron, no tag, no Workers AI call, and it works on day one for every existing brain instead of after a 90-day warm-up. I think that's both simpler and closer to the requirement.

## Smaller items

- `card--stale` is applied at `public/index.html:3209` and never defined in any stylesheet — dead class.
- The Memories browse view (`public/index.html:3360`) is a second card renderer and didn't get the stale chip, so a flagged memory looks different depending on which screen you're on.
- `computeCompoundStale` runs before `renderRecallText` drops matches over `RECALL_OUTPUT_BUDGET`, so the banner can say "3 sources are marked stale" above one visible result.
- `tagsAfterWrite` clears `stale:as-of` on append. Appending "and also X" marks the entire entry verified — touching isn't verifying. Fine if deliberate, but it should be deliberate.
- All four nightly jobs run concurrently under `ctx.waitUntil`. The staleness pass does a read-modify-write on `tags` in JS while compression `json_insert`s `rolled-up` on the same oldest-entry population. That's a lost update on overlapping rows.
- The pass doesn't exclude `auto-pattern` / `synthesized` / `rolled-up`, so its 25-row budget can be spent on rows recall can never return.

## Codebase fit

`src/ARCHITECTURE.md` defines a layer table and a module map, and the new top-level `src/staleness/` is in neither. It also straddles two layers in one directory — `heuristic.ts` is pure, `pass.ts` is domain — which is exactly the case `recall/math.ts` and `recall/rrf.ts` are individually called out for. Please register it and split or annotate accordingly.

`computeCompoundStale` is a pure function over matches but lives in `recall/render.ts`, which creates a new backwards dependency `search.ts → render.ts`. It belongs with the other pure recall helpers.

No documentation was updated — this adds two reserved tag namespaces (`volatility:`, `stale:as-of`) and a column, all of which are user-visible.

## Why the suite didn't catch any of this

Two of these are structurally invisible to the current tests, which is worth fixing regardless of this PR:

- `test/helpers/d1-mock.ts` destructures INSERT arguments **by count** and never reads the column list, so a schema mismatch like issue 3 can never fail a test.
- The pass is only ever invoked once per test, so convergence is untested by construction.

If you add a second `runStalenessPass` call and a column-name assertion in the mock, issues 1 and 3 both turn red, which is a better guarantee than either fix on its own.

---

Happy to talk through the convergence approach or the volatility precedence before you rework it — those two are genuine design calls and I don't want you guessing at what I'd accept. Everything else is mechanical.

**Comment by FaberVi** [comment]
Thanks for the thorough review — especially running the pass against real data rather than reading the diff. You're right on all four blockers, and I agree CI being green doesn't clear them.

**Plan for the rework:**

1. **Pass convergence** — I'll add a separate `staleness_checked_at` column (distinct from `updated_at`) so the nightly cursor advances without conflating "processed by cron" with "touched by user". The query will order by `staleness_checked_at ASC`, set it at the end of each row's processing, and exclude `auto-pattern` / `synthesized` / `rolled-up` from the budget. I'll add a test that calls `runStalenessPass` twice over 30+ aged entries and asserts more than 25 get processed total.

2. **`updated_at` schema** — Move to ALTER-only (no `NOT NULL` in `CREATE TABLE`), with a guarded backfill so we don't scan on every cold isolate. This also avoids the rollback trap on fresh brains.

3. **Classifier + floor precedence** — Tighten patterns; your three false positives (`email address`, `scheduled handler`, API path prose) will become regression tests. `getRecencyFloor`: canonical / importance before volatility (explicit signals beat heuristic). The pass will not overwrite an existing `volatility:` tag once a human has set it.

4. **Compound stale** — Compute from `updatedAt` age at recall assembly time (90d threshold), not from `stale:as-of`. The tag stays as a per-entry qualifier from the nightly pass; the compound warning works on day one for existing brains.

**Smaller items I'll include:** move `computeCompoundStale` out of `render.ts`; register the module in `ARCHITECTURE.md`; fix `card--stale` CSS and align the browse view; compute compound stale on visible matches after the output budget; d1-mock column-name assertion on INSERT.

**Question before I push:** Does `staleness_checked_at` as a separate column work for you, or would you prefer a reserved tag like `staleness:checked`? Happy to go either way — I lean column because it keeps tag namespaces clean and doesn't compete with compression.

I'll push the rework once you confirm.

**Comment by rahilp** [comment]
**Effort:** High - Four blockers are still open and the ranking half needs a rethink before it does anything.
**Risk:** High - A schema change with a rollback trap, plus a cron job that re-ranks memories nobody touched.
**Impact:** Medium - The as-of qualifier is worth having, but the ranking effect does not reach real recalls yet.

**What I found**

1. **Volatility tags never reach the ranker**
The recency-floor change has no effect on the normal search path. Ranking reads tags off Vectorize metadata. That metadata is written once at embed time, and the nightly pass only updates D1. An entry the pass tagged `volatility:volatile` last night still ranks on the old default floor today.
`src/recall/math.ts:74`
`src/staleness/pass.ts:50`

2. **The pass processes the same 25 entries every night**
Any brain with more than 25 aged entries stops improving after night one. The row update writes only `tags`, so nothing marks a row as done and the ordering never advances. Still open at this head.
`src/staleness/pass.ts:22`

3. **New brains and existing brains get different schemas**
Roll the Worker back to main after this ships and capture breaks permanently on any brain created under it. `CREATE TABLE` makes `updated_at` NOT NULL with no default, while the ALTER path leaves it nullable. Still open at this head.
`src/db/init.ts:5`

Review basis: 2b47d34c44f5f80b68f6848b360d162849f74b6b

---
Automated review posted on the maintainer's behalf. Worth a human read before acting on it.

**Comment by FaberVi** [comment]
Thanks for the follow-up — the Vectorize/D1 gap on ranking is a fair catch and I should have traced the full path before proposing the floor change.

**Ranking gap (new blocker #1)**

You're right: `rerankWithTimeDecay` reads tags from Vectorize metadata, which is frozen at embed time, so nightly D1 tag updates never reach the ranker. My plan:

- At rerank time, **prefer tags from the D1-hydrated match** (we already fetch `tags` from D1 for recall assembly) and pass those into `getRecencyFloor`, falling back to Vectorize metadata only when D1 tags aren't available (e.g. pre-hydration vector-only paths).
- **No re-embed on pass tag changes** — volatility classification is a D1 concern; Vectorize metadata stays content/structural.
- Add an integration-level test that seeds D1 tags (`volatility:volatile`) divergent from Vectorize metadata and asserts the reranked score reflects the D1 floor, not the stale metadata default.

If you'd rather drop the ranking effect entirely and keep `volatility:` scoped to staleness qualifiers only, I can do that instead — but hydrating from D1 at rerank seems like the smallest fix that makes the current design actually work.

**Still on track from my earlier plan (unchanged, pending your OK):**

1. **Pass convergence** — `staleness_checked_at` column (separate from `updated_at`), query orders by it ASC, set at end of each processed row, exclude `auto-pattern` / `synthesized` / `rolled-up`. Test: two `runStalenessPass` calls over 30+ aged entries, >25 processed total.
2. **`updated_at` schema** — ALTER-only, guarded backfill, no `NOT NULL` in `CREATE TABLE`.
3. **Classifier** — tighten patterns; your three false positives become regression tests; canonical/importance before volatility in `getRecencyFloor`; pass won't overwrite an existing `volatility:` tag.
4. **Compound stale** — compute from `updatedAt` age at recall assembly (90d), independent of `stale:as-of`. Tag remains the per-entry qualifier from the nightly pass.

**Smaller items:** move `computeCompoundStale` out of `render.ts`; `ARCHITECTURE.md`; `card--stale` CSS + browse view parity; compound stale on visible matches post-budget; d1-mock INSERT column-name assertion.

**#264:** aligned on your option 1 — land here first, then rebase #264.

I'll push the rework once you confirm the D1-tags-at-rerank approach (or tell me to drop ranking scope). Happy to go with `staleness_checked_at` as a column unless you prefer the reserved-tag route.

File: src/recall/search.ts [hunk 1]
```diff
@@ -200,7 +202,7 @@ export async function recallEntries(
   const allParentIds = [...seedParentIds, ...expandedScored.map(e => e.parentId)];
   const placeholders = allParentIds.map(() => "?").join(", ");
```
Comment (rahilp) [inline src/recall/search.ts:240]: This D1 hydration happens after reranking. Please also add `tags` to the existing pre-rerank candidate statistics query and pass an ID-to-tags map into `rerankWithTimeDecay`. Use D1 tags for every tag-dependent decision: floor, half-life, task and rolled-up penalties, and tag boost. Vectorize metadata should only be the fallback.

File: src/staleness/pass.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        tags = withVolatility(tags, classified);
+      }
+
+      if (shouldFlagStale(volatility)) {
+        if (!hasStaleAsOf(tags)) tags = withStaleAsOf(tags);
+      } else if (volatility === "durable") {
+        tags = withoutStaleAsOf(tags);
+      }
+
+      await env.DB.prepare(`UPDATE entries SET tags = ? WHERE id = ?`)
```
Comment (rahilp) [inline src/staleness/pass.ts]: This whole-string write races with nightly compression, which runs concurrently and also edits `tags`. Please use an atomic/CAS update or sequence the overlapping work so neither pass can erase the other’s changes.

File: src/memory/stale.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  return [...tags, STALE_AS_OF];
+}
+
+export function withoutStaleAsOf(tags: string[]): string[] {
+  return tags.filter(t => t !== STALE_AS_OF);
+}
+
+/** Strip staleness flag after a write refresh — entry was just touched. */
+export function tagsAfterWrite(tags: string[]): string[] {
+  return withoutStaleAsOf(tags);
```
Comment (rahilp) [inline src/memory/stale.ts]: Clearing only `stale:as-of` leaves a system-derived `volatility:` tag attached after content changes. If existing volatility is never overwritten, a replace can freeze the old classification forever. Please clear and reclassify system volatility on content-changing writes, or distinguish system-derived values from user-set tags.

**Review by rahilp — COMMENTED** [review]
Yes, use `staleness_checked_at` as a nullable, ALTER-only column. It is operational cursor state, not semantic memory, so keep it out of tags and exports. Advance it for every inspected row, including `null` classifications, and add the multi-run convergence test.

D1-authoritative tags at rerank is also the right direction, subject to the inline details below. With those constraints, the plan works for me. Please push the rework and rebase on current `main`, then I’ll rerun the review.

**Comment by FaberVi** [comment]
Rework pushed per review 4840433632 — rebased on current `main`, ready for re-review.

**Changes in `98b7696`:**
- `staleness_checked_at` cursor (ALTER-only) + guarded `updated_at` backfill
- Staleness pass: COALESCE cursor ordering, system-tag exclusions, CAS tag updates, no overwrite of existing `volatility:*`, cursor advanced even on `null` classification
- Rerank: D1 tags authoritative pre-rerank; canonical/importance floor before volatility
- `tagsAfterWrite`: clears `stale:as-of` and system `volatility:*` on content-changing writes
- Compound stale from `updatedAt` age ≥90d (after output budget in render)
- Classifier tightened + false-positive regressions; convergence test (2× pass >25 inspected)
- UI: `card--stale` CSS + browse-view stale chip; `src/staleness/` in ARCHITECTURE.md

`npx vitest run` — 756/756 green.

File: src/recall/search.ts (hunk 1 above)
Comment (FaberVi) [reply src/recall/search.ts:240]: Fixed in `98b7696` — the pre-rerank stats query now selects `tags`, builds a `Map<id, tags[]>`, and passes it into `rerankWithTimeDecay`. Tag-dependent decisions (floor, half-life, rolled-up penalty, tag boost) use D1 first, Vectorize metadata as fallback. Unit test: D1 `volatility:volatile` vs Vectorize `status:canonical` → score reflects the D1 floor.

File: src/staleness/pass.ts (hunk 2 above)
Comment (FaberVi) [reply src/staleness/pass.ts]: Fixed in `98b7696` — tag updates use CAS: read tags, merge in JS, then `UPDATE ... SET tags = ?, staleness_checked_at = ? WHERE id = ? AND tags = ?` with up to 3 retries. Avoids lost updates when compression `json_insert`s on the same row concurrently.

File: src/memory/stale.ts (hunk 3 above)
Comment (FaberVi) [reply src/memory/stale.ts]: Fixed in `98b7696` — `tagsAfterWrite` now calls `withoutVolatility(withoutStaleAsOf(tags))`, so content-changing writes clear both `stale:as-of` and system `volatility:*`. The nightly pass still won't overwrite an existing `volatility:` tag (user-set); after a write the tag is cleared and reclassified on the next pass. Covered in `capture-entry`, `append`, and integration tests.

**Comment by rahilp** [comment]
**Effort:** Medium - The blockers are cleared. What is left is tuning when the warning fires. (was High)
**Risk:** Low - The schema is additive and nullable now, and the pass advances a cursor. (was High)
**Impact:** Medium - The as-of qualifier works end to end now. The compound warning needs a filter.

**Since my last review**

Fixed: all three. D1 tags now override Vectorize metadata in rerank, so volatility actually moves the floor. The pass writes `staleness_checked_at` and orders by it, so it walks past the first 25. `updated_at` is ALTER-only with a guarded backfill, and the capture INSERT sets it.

1. **The staleness banner will fire on almost every recall**
It counts any match untouched for 90 days, whatever its volatility. A birthday from 2019 counts the same as a stale job title. Two of those in a top-5 recall are enough to trigger it. On any brain with real history the banner shows up constantly and stops meaning anything. Filter on `stale:as-of` rather than age alone.
`src/recall/compound-stale.ts:7`

2. **The MCP count includes matches it never printed**
`renderRecallText` recomputes the warning on the matches that survived the output budget. The MCP caller passes the pre-budget value and that one wins. So the text can warn about four stale sources while showing two.
`src/mcp/server.ts:214`

3. **An edit wipes a volatility tag a person set**
The pass is careful not to overwrite a human `volatility:` tag. But `tagsAfterWrite` strips every one of them on any append, merge or update. The next classify then guesses again from the heuristic.
`src/memory/stale.ts:20`

Review basis: 98b7696b75c84951b12f821208549afd6520e390

---
Automated review posted on the maintainer's behalf. Worth a human read before acting on it.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/recall/compound-stale.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import type { RecallMatch, CompoundStaleSignal } from "./types";
+
+export const COMPOUND_STALE_AGE_MS = 90 * 24 * 60 * 60 * 1000;
+
+/** Warn when multiple recalled sources have not been touched in 90+ days. */
+export function computeCompoundStale(matches: RecallMatch[], now = Date.now()): CompoundStaleSignal | undefined {
+  const aged = matches.filter(m => now - m.updatedAt >= COMPOUND_STALE_AGE_MS);
```
Comment (rahilp) [inline src/recall/compound-stale.ts]: This counts any match untouched for 90 days, whatever its volatility. The backfill sets `updated_at` to `created_at` for every existing row, so on a brain with real history almost everything qualifies from day one, and two in a recall is enough to fire the banner.

The result is a warning prepended to nearly every recall, on both the MCP and HTTP paths. A warning that always fires stops carrying information, and on the MCP path it also pushes the assistant to hedge on every memory backed answer, which is worse than not having the warning at all.

Filtering on the `stale:as-of` flag rather than raw age fixes it, and `staleAsOf` is already on the match from `search.ts`:

```ts
const aged = matches.filter(m => m.staleAsOf && now - m.updatedAt >= COMPOUND_STALE_AGE_MS);
```

The `computeCompoundStale` cases in `test/unit/render-stale.test.ts` will need their fixtures updated to set `staleAsOf`.

**Review by rahilp — COMMENTED** [review]
The rework cleared all three blockers. I checked rather than taking it on trust: `updated_at` is ALTER only and nullable with the guarded backfill so the rollback trap is gone, the `staleness_checked_at` cursor means the pass walks past the first 25, and the rerank really does prefer D1 tags over Vectorize metadata with canonical and importance ahead of volatility. That was a good fix on all three.

Two mechanical things before this can land, plus one line I would like changed.

The branch is 93 commits behind and conflicts with main in 10 files, including `src/recall/search.ts`, `src/recall/math.ts`, `src/recall/render.ts` and `src/mcp/server.ts`. #269 and #271 rewrote a lot of that same ranking code, so these are semantic conflicts rather than textual ones and worth working through carefully.

CI has also never run on this head. The only runs on the branch are from 31 July against the old base. Your 756 passing matches what I get locally, it just has not been checked against the merged result.

The one change is inline below. Everything else I found is minor and is now in #272.

**Comment by FaberVi** [comment]
Rebased onto current \main\ (includes #269/#271) and pushed \74329de\.

**Review 4847545503:**
- \computeCompoundStale now filters on \staleAsOf (not raw age alone), so durable old facts no longer trigger the compound banner.
- Full suite green: \
px vitest run\ — 1018 passing.

Ready for re-review / CI on the rebased head.

**Comment by FaberVi** [comment]
CI typecheck fix pushed in \887d792\ — \
ender-recall.test.ts\ helper now supplies required \updatedAt\ on \RecallMatch\. Local: typecheck + 1018 tests green.

**Review by rahilp — APPROVED** [review]
Approving. The three blockers from the July review are genuinely resolved, and I verified each rather than taking the diff on trust: `updated_at` is ALTER-only and nullable, the `staleness_checked_at` cursor makes the pass walk past the first 25 (measured 25 / 50 / 60 over three runs against a real Workers runtime), and the rerank prefers D1 tags with canonical and importance ahead of volatility. The `staleAsOf` filter on compound staleness is in. The classifier rewrite is a big improvement — over the same 20 realistic memories that flagged 3 of 14 neutral notes before, it now flags none, and all six genuine cases classify correctly.

I've pushed a follow-up commit rather than sending this back again, since the remaining items were mine to find and mostly concern free-plan cost rather than the feature itself. Summary of what changed and why:

**The `updated_at` backfill is gone.** Every reader already coalesces the column to `created_at`, so a NULL and a written value are indistinguishable downstream — which makes the backfill a no-op that still costs one row written per entry. On a 50k-entry brain that was half a day's free-plan write budget in one statement, and exceeding D1's daily write cap fails every query account-wide until 00:00 UTC. `initializeDatabase` now issues zero queries against `entries` on every path. A test walks `src/` and fails if any reader stops coalescing, so the invariant that makes this safe is enforced rather than assumed.

**`initializeDatabase` rejects instead of swallowing.** It is memoised per isolate to stop the four nightly jobs each paying for the same dozen DDL statements. That memo has to key on success — a resolved promise was caching a schema that may never have been applied, leaving every later caller in that isolate doing nothing against an unmigrated database. Callers now handle the rejection so a failed init retries.

**The CAS guards `content` as well as `tags`.** Classification derives from content, and for an entry carrying neither reserved tag the mutation is a no-op on tags — so a concurrent rewrite left the guard satisfied and committed a verdict about content that no longer existed. It did not self-correct either, because the same write pushed `updated_at` past the cutoff. Reproduced as a memory rewritten into a durable fact being tagged stale milliseconds later, then staying that way for 90 days.

**Cost.** One nightly run drops from 68 D1 statements to 41, and the pass itself from 64 to 38 by reusing the tags the candidate query already selected. `compressTag` now checks for a system tag before `resolveConfig`, so a skipped tag no longer costs a KV read.

**Dashboard.** `card--stale` used a hardcoded light background, which put tertiary text below WCAG AA in dark mode. No tint satisfies both themes, so it now uses a dashed border in a new semantic `--warn` colour, matching how `card--synthesized` and `card--rolled-up` already work.

The branch is merged up to current `main` and CI is green on this head — worth noting the earlier run was red on the typecheck, so this is the first end-to-end verification against current `main`.

One thing that is **not** yours to fix, recorded so it does not look like an oversight: at a realistic brain size the whole nightly cron issues around 316 D1 statements, dominated by nightly compression rather than by this feature. Roughly 350 of those are a single per-entry `UPDATE` loop in `digest.ts` that should be batched. That predates this PR and is tracked in #278; this work reduces its own contribution to it.

Thanks for sticking with this through several rounds — the convergence cursor and the classifier narrowing were the right calls and they made the rest tractable.
